packages/eslint-plugin/src/rules/restrict-template-expressions.ts

```diff
@@ -44,29 +44,36 @@ export default util.createRule<Options, MessageId>({
     const service = util.getParserServices(context);
     const typeChecker = service.program.getTypeChecker();
 
-    type BaseType =
-      | 'string'
-      | 'number'
-      | 'bigint'
-      | 'boolean'
-      | 'null'
-      | 'undefined'
-      | 'other';
-
-    const allowedTypes: BaseType[] = [
-      'string',
-      ...(options.allowNumber ? (['number', 'bigint'] as const) : []),
-      ...(options.allowBoolean ? (['boolean'] as const) : []),
-      ...(options.allowNullable ? (['null', 'undefined'] as const) : []),
-    ];
-
-    function isAllowedType(types: BaseType[]): boolean {
-      for (const type of types) {
-        if (!allowedTypes.includes(type)) {
-          return false;
-        }
+    function isUnderlyingTypePrimitive(type: ts.Type): boolean {
+      if (util.isTypeFlagSet(type, ts.TypeFlags.StringLike)) {
+        return true;
+      }
+
+      if (
+        util.isTypeFlagSet(
+          type,
+          ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike,
+        ) &&
+        options.allowNumber
```

Comment:
nit - do the option check first to save some cycles
```suggestion
        options.allowNumber &&
        util.isTypeFlagSet(
          type,
          ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike,
        )
```